esp/services/esdl_svc_engine/esdl_binding.cpp

```diff
@@ -860,6 +860,29 @@ void EsdlServiceImpl::handleFinalRequest(IEspContext &context,
 
             soapmsg.append(req.str());
             soapmsg.append("</").append(tgtQueryName).append("></soap:Body>");
+
+            auto cfgGateways = tgtcfg->queryBranch("Gateways");
+            if (cfgGateways)
+            {
+                auto baseXpath = cfgGateways->queryProp("@transformTarget");
+                if (!isEmptyString(baseXpath))
+                {
+                    Owned<IPTree> gws = createPTree("gateways", 0);
+                    VStringBuffer tmpSoap("%s</soap:Envelope>", soapmsg.str());
+                    Owned<IPTree> soapTree = createPTreeFromXMLString(tmpSoap);
+                    StringBuffer xpath(baseXpath);
+
+                    EsdlBindingImpl::transformGatewaysConfig(tgtcfg, gws, "row");
+                    xpath.replaceString("{$query}", tgtQueryName);
+                    xpath.replaceString("{$method}", mthdef.queryMethodName());
+                    xpath.replaceString("{$service}", srvdef.queryName());
+                    xpath.replaceString("{$request}", mthdef.queryRequestType());
+                    mergePTree(ensurePTree(soapTree, xpath), gws);
+                    toXML(soapTree, tmpSoap.clear());
+                    tmpSoap.replaceString("</soap:Envelope>", "").trim();
```

Comment:
replaceString seems kind of expensive here.  Can't you just remove from the end?  toXML is pretty predictable in output format.